### ISSUE TITLE
Absolute directory paths and existence check

### ISSUE THREAD

**Issue by PragTob** [cucumber/aruba#478]
First thanks for aruba :tada: 

## Summary

I upgraded aruba in simplecov from 0.7 to 0.14.2. One spec started failing, basically it configures a different directory for simplecovs output as:

      SimpleCov.start do
        coverage_dir '/tmp/test/simplecov'
      end

The test asserting that the directory exists started failing, and I fear that it also won't find the files I'm checking for.

It is imo because the method:

```ruby
      def directory?(file)
        Aruba.platform.directory? expand_path(file)
      end
```

calls out to `expand_path` which gets the path wrong:

```
[1] pry(#<Object>)> file
=> "/tmp/test/simplecov"
[2] pry(#<Object>)> expand_path(file)
=> "/home/tobi/github/simplecov/tmp/aruba/project/tmp/test/simplecov"
```
So `/tmp/test/simplecov` exists (and we created it using `File.mkdir_p`) but the path aruba thinks that it should be does not exist, hence the test fails ([this is the test](https://github.com/colszowka/simplecov/blob/master/features/config_coverage_dir.feature#L22-L33))

## Expected Behavior

I'd expect aruba to find directories/files created by my application in the tests. 

## Current Behavior

Aruba looks in its working directory not the actual absolute path.

## Possible Solution

One could check if the path is absolute and then just not expand it and everything should work fine.

## Steps to Reproduce (for bugs)

[scenario failing when updating aruba](https://github.com/colszowka/simplecov/blob/master/features/config_coverage_dir.feature#L22-L33)

Basically just running `directory?('/some/absolute/path')` seems to fail for me.

## Context & Motivation

People might wanna put their simplecov coverage reports in an absolute path to gather them or whatever. I wanna make sure they can do that.

## Your Environment

* Version used: 0.14.2
* Operating System and version:
  * Linux Mint 18.2
  * `uname -a`: Linux speedy 4.10.0-32-generic #36~16.04.1-Ubuntu SMP Wed Aug 9 09:19:02 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
* Link to your project: https://github.com/colszowka/simplecov
* Ruby: 2.4

**Comment by PragTob** [comment]
A quick fix that works for me (happy to PR but I think there's more involved, still happy to PR it):

```
module Aruba
  module Api
    module Core
      def expand_path(file_name, dir_string = nil)
        # only line added, don't ask why but I couldn't get alias_method to work...
        return file_name if absolute?(file_name)
        # normal method
      end
    end
  end
end
```

**Comment by maxmeyer** [comment]
@PragTob Thanks for taking time and open issue for this! The difference between those two tests are relative vs. absolute paths. Correct? And this is you want to make sure works?

The idea behind all methods of aruba is to have an isolated directory where everything happens. Personally I don't think we should make an exception to this.

I see your point that we broke the API and this should not happen without a major version change. I see the following options:

1. Make aruba 0.14 work again like the older ones
2. @PragTob You change your tests a bit to work around the problem - you could use `~` instead https://github.com/cucumber/aruba/blob/master/lib/aruba/api/core.rb#L112 / https://github.com/cucumber/aruba/blob/51a034a6380f9686c045cdbb7639fcb922046959/features/04_aruba_api/core/expand_path.feature#L57

Eventually I would like to consider only files within the working directory of aruba (`tmp/aruba`) as of aruba 1.0.0. That's the only directory we have control of.  Bringing back compatibility makes sense as well. So adding support for aruba < 1.0.0 is not a problem from my side. Additionally I think it makes sense to have a discussion about how to handle tests with absolute paths in a separate issue. As this might be a more common problem.

@mvz @olleolleolle @junaruga What do you think?

**Comment by mvz** [comment]
Digging into that test, I see that the Aruba step used is `And a directory named "#{coverage_dir}" should exist`.

From a user perspective, it makes sense that, if an absolute path is specified there, Aruba should just use that path as-is.

On the other hand, I think tests should not touch any files outside the `tmp/aruba` directory, and, e.g., writing to or even reading from files in the user's home directory is generally a bug. `/tmp` is usually safe, but it would be tedious and error-prone to add heuristics for that.

Combining those two considerations, my proposal would be:

* Keep absolute paths as specified by the user (so restoring to pre-0.14 behavior).
* Complain about it loudly.
* Provide an annotation so the user can say "Yes, I'm using an absolute path here, but I know what I'm doing and I'm not clobbering the user's files", and Aruba will stop complaining.

**Comment by PragTob** [comment]
@mvz I like the proposal.

I like that aruba looks to run in "its own world" which is great. Sometimes I think the absolute paths are necessary though (for testing) - I don't expect lots of (simplecov) users to use absolute paths like this but I can easily imagine that on a CI with multiple workers (and potentially multiple users) you want to have a directory where you store all their coverage information. Which is why I still wanna have that test to see that nothing breaks :)

**Comment by mvz** [comment]
I've made a separate ticket for the first of my three points. That's the one I think should go into a potential 0.14.3 release.

**Comment by mvz** [comment]
I'll try to pick this up soon. @maxmeyer do you agree with my three-point plan?

**Comment by maxmeyer** [comment]
@mvz Sounds reasonable. Do you plan to add a config option for that?

**Comment by mvz** [comment]
@maxmeyer I was considering a cucumber annotation, but a config option (or both) may be better since that will work with rspec.

**Comment by maxmeyer** [comment]
I think using a config option seems to be the best options.

**Comment by matthutchinson** [comment]
I came across this issue too, upgrading [lolcommits](https://github.com/mroth/lolcommits) to use the latest Aruba, from a much older version.  

We have a few features that require running outside `tmp/aruba` (we need to be outside the projects dir altogether to avoid being in a git repo).  For these tests we [used to](https://github.com/mroth/lolcommits/blob/master/features/support/env.rb#L51) make a new tmp dir, simply set `@dirs`.

I tried patching aruba/core.rb as @PragTob suggested but couldn't get that to work, so instead, I hacked this up which seems to work for now;

```
# in support/env.rb
Before('@no-repo-dir') do
  @original_root = aruba.root_directory.pop
  @original_dir  = aruba.current_directory.pop
  @working_dir   = Dir.mktmpdir
  aruba.current_directory << @working_dir
end

After('@no-repo-dir') do
  aruba.current_directory.pop
  aruba.root_directory << @original_root
  aruba.current_directory << @original_dir

  FileUtils.rm_rf(@working_dir)
end

# then use @no-repo-dir in features that should run within this tmp dir.
```

It would be nice if this were a feature Aruba could provide, run tests inside a new tmp dir, then clean it up afterwards.

**Comment by mvz** [comment]
> we need to be outside the projects dir altogether to avoid being in a git repo

That's a clear use case if ever I saw one! Thanks, @matthutchinson.

**Comment by mvz** [comment]
@matthutchinson is your case solved with the latest 0.14.x version?

**Comment by matthutchinson** [comment]
will try that now and let you know

**Comment by matthutchinson** [comment]
I can't see the changes in [this PR](https://github.com/cucumber/aruba/pull/540) in `v0.4.15`

**Comment by mvz** [comment]
That's correct, since that's the PR for the upcoming 1.0 release. You should see the changes from #539 there though.

**Comment by matthutchinson** [comment]
Sorry, this doesn't work for me, I'm trying this;

```
Given(/^I am not in a git repo$/) do
  tmp_dir = Dir.mktmpdir
  steps %(
    And I cd to "#{tmp_dir}"
    )
end
```

To run a commands in a tmp directory.. But `expand_path('.')`  from the `cd` method  in `api/core.rb` will always prefix this with `aruba.root_directory, aruba.current_directory`. Here's a debugging session to illustrate , the tmp dir is `/tmp/some_dir5`:

<img width="850" alt="screen shot 2018-04-29 at 20 59 47" src="https://user-images.githubusercontent.com/2095/39410448-473904dc-4bf0-11e8-8e57-2bed4dc5e25a.png">

Should I be doing something different? Perhaps I shouldn't use `cd` for this and instead reconfigure Aruba in a Before block (with cuke tag).
